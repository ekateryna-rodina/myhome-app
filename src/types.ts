// import { gql } from "@apollo/client";
// import { makeExecutableSchema } from "apollo-server-micro";
// import { createContext } from "./context";
// const prisma = createContext();
// export const typeDefs = gql`
//   type Property {
//     id: ID
//     title: String
//     city: String
//     country: String
//     baths: Int
//     beds: Int
//     size: Int
//     photo: String
//   }
//   type Query {
//     properties: [Property]
//   }
// `;

// export const resolvers = {
//   Query: {
//     properties: (_parent, args, context) => {
//       return prisma.property.findMany();
//     },
//   },
// };

// export const schema = makeExecutableSchema({
//   typeDefs,
//   resolvers,
// });
