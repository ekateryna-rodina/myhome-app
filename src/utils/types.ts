import { PropertyType } from "./enums";

export type Location = {
  id: number;
  city: string;
  country: string;
  zip?: string;
};
export type Listing = {
  id: number;
  title: string;
  locationId: number;
  location: Location;
  baths: number;
  beds: number;
  size: number;
  photo: string;
  for: string;
};

type AdditionalFilters =
  | "isPetsFriendly"
  | "isFurnished"
  | "isParkingAccessible"
  | "isWithKitchen"
  | "isWithAirCondition"
  | "isWithLaundry"
  | "isWithBabyBed"
  | "isNearbyBeach"
  | "isWithOfficeZone"
  | "isWithSmokingZone"
  | "isWithWifi"
  | "isWithBreakfast"
  | "isWithFireplace";

export interface Filter {
  propertyTypes: Record<PropertyType, boolean>;
  additional?: { [key in AdditionalFilters]: boolean };
  priceRange?: number[];
  sizeRange?: number[];
  bathrooms?: number;
  bedrooms?: number;
}
