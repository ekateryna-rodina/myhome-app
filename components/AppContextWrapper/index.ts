export { default as AppContextWrapper } from "./AppContextWrapper";
