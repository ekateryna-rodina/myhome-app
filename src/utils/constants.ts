import { gql } from "@apollo/client";

export const MAX_FILTER_PRICE = 5000;
export const MIN_FILTER_PRICE = 20;

export const MIN_FILTER_SIZE = 100;
export const MAX_FILTER_SIZE = 5000;

export const FILTER_BEDROOM = 2;
export const FILTER_BATHROOM = 1;

// ten is replaced by another label in a view
export const DEFAULT_ROOMS_NUMBER_LIST = [1, 2, 3, 10];
export const FILTER_ADDITIONAL_MAP = {};

export const GET_PROPERTIES_QUERY = gql`
  query properties($locationId: Int, $filter: String) {
    properties(locationId: $locationId, filter: $filter) {
      id
      title
      beds
      baths
      size
      photo
      locationId
      location {
        city
        country
      }
    }
  }
`;
