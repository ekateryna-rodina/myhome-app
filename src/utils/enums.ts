enum Unit {
  USD = "$",
  SQFT = "sqft",
}

enum Icons {
  Glass = "glass",
  Filter = "filter",
  Office = "office",
  House = "house",
  Apartment = "apartment",
  Landplot = "landplot",
  Message = "message",
  Notification = "notification",
  Bed = "Beds",
  Bath = "Baths",
  Size = "sqft",
  Sad = "",
  Close = "",
  Visit = "",
  Favourite = "",
}

enum ServiceType {
  BUY = "Buy",
  RENT = "Rent",
}

enum PropertyType {
  Apartment = "Apartment",
  House = "House",
  Office = "Office",
  Landplot = "Landplot",
}

enum PropertyFor {
  RENT = "Rent",
  BUY = "Buy",
}

export { Unit, Icons, ServiceType, PropertyType, PropertyFor };
