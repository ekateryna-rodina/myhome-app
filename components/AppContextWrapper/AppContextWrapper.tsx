import React, { useCallback, useState } from "react";
import { initialFilter } from "../../src/utils/constants";
import { Filter, Listing, Location } from "../../src/utils/types";
type ContextPropsType = {
  locations: Location[];
  properties: Listing[] | null;
  isFilterOpen: boolean | null;
  loading: boolean;
  filter: Filter;
  selectedLocationId: number;
  focusItemListId: number | null;
  resultsChanged: boolean;
  handleFilter: (filter: Filter) => void;
  handleLoading: (value: boolean) => void;
  handleLocations: (locations: Location[]) => void;
  handleIsFilterOpen: (value: boolean) => void;
  handleProperties: (properties: Listing[]) => void;
  handleSelectedLocationId: (locationId: number) => void;
  handleFocusItemListId: (id: number) => void;
  handleResultsChanged: (isChanged: boolean) => void;
};

const initialState: ContextPropsType = {
  locations: [],
  properties: null,
  isFilterOpen: false,
  loading: false,
  filter: initialFilter,
  selectedLocationId: 0,
  focusItemListId: null,
  resultsChanged: false,
  handleFilter: (filter: Filter) => {},
  handleLoading: (value: boolean) => {},
  handleIsFilterOpen: (value: boolean) => {},
  handleLocations: (locations: Location[]) => {},
  handleProperties: (properties: Listing[]) => {},
  handleSelectedLocationId: (id: number) => {},
  handleFocusItemListId: (id: number) => {},
  handleResultsChanged: (isChanged: boolean) => {},
};
export const AppContext = React.createContext(initialState);
type AppContextProviderProps = { locations: Location[] };
const AppContextWrapper: React.FC<AppContextProviderProps> = ({
  children,
  locations: injectedLocations,
}) => {
  const [isFilterOpen, setIsFilterOpen] = useState<boolean | null>(null);
  const [locations, setLocations] = useState<Location[]>(
    injectedLocations || []
  );
  const [resultsChanged, setResultsChanged] = useState<boolean>(false);
  const [properties, setProperties] = useState<Listing[] | null>(null);
  const [loading, setLoading] = useState<boolean>(false);
  const [filter, setFilter] = useState<Filter>(initialFilter);
  const [focusItemListId, setFocusItemListId] = useState<number | null>(null);
  const [selectedLocationId, setSelectedLocationId] = useState<number>(0);
  const handleLocations = (locations: Location[]) => {
    setLocations(locations);
  };
  const handleIsFilterOpen = (open: boolean) => {
    setIsFilterOpen(open);
  };
  const handleProperties = (listings: Listing[]) => {
    setProperties(listings);
  };
  const handleLoading = (show: boolean) => {
    setLoading(show);
  };
  const handleFilter = useCallback(
    (filter: Filter) => {
      setFilter(filter);
    },
    [isFilterOpen]
  );
  const handleSelectedLocationId = (locationId: number) => {
    setSelectedLocationId(locationId);
  };
  const handleFocusItemListId = (id: number) => {
    setFocusItemListId(id);
  };
  const handleResultsChanged = (isChanged: boolean) => {
    setResultsChanged(isChanged);
  };
  const contextProps = {
    selectedLocationId,
    handleSelectedLocationId,
    locations,
    handleLocations,
    isFilterOpen,
    handleIsFilterOpen,
    properties,
    handleProperties,
    loading,
    handleLoading,
    filter,
    handleFilter,
    focusItemListId,
    handleFocusItemListId,
    resultsChanged,
    handleResultsChanged,
  };
  return (
    <AppContext.Provider value={contextProps}>{children}</AppContext.Provider>
  );
};

export default AppContextWrapper;
