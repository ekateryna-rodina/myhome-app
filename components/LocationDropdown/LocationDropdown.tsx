import React, { useContext, useEffect, useState } from "react";
import styled from "styled-components";
import HeaderButton from "../../components/HeaderButton.style";
import { Icons } from "../../src/utils/enums";
import useAutocomplete from "../../src/utils/hooks/useAutocomplete";
import { Location } from "../../src/utils/types";
import { respondTo } from "../../src/utils/_respondTo";
import { FilterContext } from "../FilterProviderWrapper/FilterProviderWrapper";

const Container = styled.div`
  position: relative;
  flex: 4;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding-right: 0.3rem;
  margin-left: 0;
  ${respondTo.tablet`
  margin-left: 1rem;
  `}
  ${respondTo.laptopAndDesktop`
  margin-left: 1rem;
  `}
`;
const LocationInput = styled.input`
  height: 90%;
  width: 100%;
  font-size: 0.9rem;
  border: 0;
  outline: 0;
  color: ${(props) => props.theme.gray};
`;

const MarkerIcon = styled.div`
  color: ${(props) => props.theme.gray};
  margin-right: 0.3rem;
  :after {
    font-family: "Font Awesome 5 Free";
    font-weight: 900;
    content: "\f3c5";
  }
`;
const SearchButtonContainer = styled.div`
  position: absolute;
  top: 50%;
  right: 0.35rem;
  transform: translateY(-50%);
`;

const LocationsOptionsContainer = styled.div<{ show: boolean }>`
  position: absolute;
  left: 0;
  top: calc(100% + 0.6rem);
  width: 100%;
  height: 10rem;
  background: #fff;
  border-radius: 0.5rem;
  box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.3);
  opacity: ${({ show }) => (show ? "1" : "0")};
`;

const OptionsList = styled.ul`
  list-style: none;
  padding: 0;
`;
const LocationDropdown = () => {
  const { locations } = useContext(FilterContext);

  const [state, setState] = useState<{
    value: string;
    filteredOptions: Location[];
    activeOptions: number[];
  }>({
    value: "",
    filteredOptions: [],
    activeOptions: [],
  });
  const filteredOptions = useAutocomplete(state.value, locations);
  useEffect(() => {
    setState({ ...state, filteredOptions });
    //eslint-disable-next-line
  }, [filteredOptions]);
  const onKeyDownHandler = () => {};
  const onFilteredClickHandler = (id: number) => {
    const { city, country } = filteredOptions.filter((o) => o.id === id)[0];
    setState({
      ...state,
      activeOptions: [...state.activeOptions, id],
      value: `${city}, ${country}`,
    });
  };
  return (
    <Container data-testid="locationDropdownTestId">
      <MarkerIcon />
      <LocationInput
        value={state.value}
        placeholder="Where should I search?"
        onChange={(e) => setState({ ...state, value: e.currentTarget.value })}
        onKeyDown={onKeyDownHandler}
        data-testid="locationInputTestId"
      />
      <SearchButtonContainer>
        <HeaderButton icon={Icons.Glass} handler={() => console.log()} />
      </SearchButtonContainer>
      <LocationsOptionsContainer
        data-testid="locationsOptionsContainerTestId"
        show={filteredOptions.length > 0}
      >
        <OptionsList data-testid="locationsListTestId">
          {filteredOptions.length > 0 ? (
            filteredOptions.map((o) => (
              <li key={o.id} onClick={onFilteredClickHandler.bind(null, o.id)}>
                {o.city}, {o.country}
              </li>
            ))
          ) : (
            <li></li>
          )}
        </OptionsList>
      </LocationsOptionsContainer>
    </Container>
  );
};

export default LocationDropdown;
