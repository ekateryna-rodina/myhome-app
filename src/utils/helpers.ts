import {
  MAX_FILTER_PRICE,
  MAX_FILTER_SIZE,
  MIN_FILTER_PRICE,
  MIN_FILTER_SIZE,
} from "./constants";
import { PropertyFor } from "./enums";
import {
  AdditionalFilterKeys,
  AdditionalFiltersStringMap,
  Filter,
} from "./types";

const getRoomsNumber = (original: number[]) => {
  // update based on real data in future
  const max = 20;
  const mask = 10;
  const startNumber = 4;
  if (original.indexOf(mask) === -1) return original;
  const roomsNumberAfterMaskRemoved = [
    ...original.filter((i) => i !== mask),
  ].concat(Array.from(Array(max).keys()).map((n) => n + startNumber));
  return roomsNumberAfterMaskRemoved;
};

export const composeWhere = (locationId: number, filterQuery: string) => {
  const queryList = [];
  if (locationId) {
    queryList.push({ locationId: { equals: locationId } });
  }
  if (!filterQuery)
    return {
      AND: queryList,
    };
  const filter: Filter = JSON.parse(filterQuery);
  // PROPERTY TYPES
  const propertyTypeValues = Object.values(filter.propertyTypes);
  const selectAllTypes =
    propertyTypeValues.every((value) => !value) ||
    propertyTypeValues.every((value) => value);
  if (!selectAllTypes) {
    const showPropertTypes: string[] = Object.entries(
      filter.propertyTypes
    ).reduce((result: string[], entry: [string, boolean]) => {
      if (entry[1]) {
        result.push(entry[0]);
      }
      return result;
    }, []);

    queryList.push({ type: { in: showPropertTypes } });
  }
  // PRICE RANGE
  if (
    filter.priceRange[0] !== MIN_FILTER_PRICE ||
    filter.priceRange[1] !== MAX_FILTER_PRICE
  ) {
    queryList.push(
      ...[
        { price: { gte: filter.priceRange[0] } },
        { price: { lte: filter.priceRange[1] } },
      ]
    );
  }
  // SIZE RANGE
  if (
    filter.sizeRange[0] !== MIN_FILTER_SIZE ||
    filter.sizeRange[1] !== MAX_FILTER_SIZE
  ) {
    queryList.push(
      ...[
        { size: { gte: filter.sizeRange[0] } },
        { size: { lte: filter.sizeRange[1] } },
      ]
    );
  }

  // BEDROOMS
  queryList.push({ beds: { in: getRoomsNumber(filter.bedrooms) } });

  // BATHDROOMS
  queryList.push({ baths: { in: getRoomsNumber(filter.bathrooms) } });

  // ADDITIONAL
  const keys = Object.keys(AdditionalFiltersStringMap);
  keys.forEach((key) => {
    const isImportant = filter.additional[key as AdditionalFilterKeys];
    if (isImportant) {
      queryList.push({
        [key as AdditionalFilterKeys]: {
          equals: true,
        },
      });
    }
  });

  // FOR
  // fix case toUpper case
  queryList.push({
    for: {
      equals: filter.for.toUpperCase(),
    },
  });

  // MAP COORDINATES
  const { minLat, maxLat, minLng, maxLng } = filter.mapCoordinates;
  if (minLng || maxLng || minLat || maxLat) {
    queryList.push(
      ...[
        { lat: { gte: minLat } },
        { lat: { lte: maxLat } },
        { long: { gte: minLng } },
        { long: { lte: maxLng } },
      ]
    );
  }

  return { AND: queryList };
};

export function enumToStringArray<Type extends PropertyFor>(
  arg: Type
): string[] {
  let result: string[] = [];
  const isString = (val: number | string) => isNaN(+val);
  result = Object.keys(arg)
    .filter(isString)
    .map((k: string) => arg[k as any]);
  return result;
}
