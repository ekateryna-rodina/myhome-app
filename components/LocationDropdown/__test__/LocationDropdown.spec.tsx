import "@testing-library/jest-dom/extend-expect";
import {
  cleanup,
  fireEvent,
  render,
  screen,
  within,
} from "@testing-library/react";
import React from "react";
import { Location } from "../../../src/utils/types";
import { FilterProviderWrapper } from "../../AppContextWrapper";
import LocationDropdown from "../LocationDropdown";
jest.mock("../../../assets/bed.svg", () => () => null);
jest.mock("../../../assets/bell.svg", () => () => null);
jest.mock("../../../assets/building.svg", () => () => null);
jest.mock("../../../assets/email.svg", () => () => null);
jest.mock("../../../assets/favourite.svg", () => () => null);
jest.mock("../../../assets/filter.svg", () => () => null);
jest.mock("../../../assets/house-design.svg", () => () => null);
jest.mock("../../../assets/house.svg", () => () => null);
jest.mock("../../../assets/land.svg", () => () => null);
jest.mock("../../../assets/loupe.svg", () => () => null);
jest.mock("../../../assets/office-building.svg", () => () => null);
jest.mock("../../../assets/toilet.svg", () => () => null);
jest.mock("@atlaskit/css-reset/dist/bundle.css", () => () => null);
afterEach(cleanup);
const locations: Location[] = [
  { id: 1, city: "Philadelphia", country: "USA" },
  { id: 1, city: "Boston", country: "USA" },
  { id: 1, city: "New York", country: "USA" },
];
const setup = () => {
  render(
    <FilterProviderWrapper locations={locations}>
      <LocationDropdown />
    </FilterProviderWrapper>
  );
  const input = screen.getByTestId("locationInputTestId");
  return input;
};
test("location dropdown is rendered", async () => {
  render(<LocationDropdown />);
  expect(screen.getByTestId("locationDropdownTestId")).toBeInTheDocument();
});
test("options are hidden on initial render", async () => {
  render(<LocationDropdown />);
  expect(
    screen.getByTestId("locationsOptionsContainerTestId")
  ).toBeInTheDocument();
  expect(screen.getByTestId("locationsOptionsContainerTestId")).toHaveStyle(
    "opacity: 0"
  );
});
test("options are shown when user enters keyword which is present in initial set of location data", async () => {
  const input = setup();
  fireEvent.change(input, { target: { value: "bo" } });
  const locationsOptions = screen.getAllByTestId(
    "locationsOptionsContainerTestId"
  )[0];
  expect(locationsOptions).toHaveStyle("opacity: 1");
});

test("options are shown when user enters keyword which is present in initial set of location data", async () => {
  const input = setup();
  fireEvent.change(input, { target: { value: "notacity" } });
  const locationsOptions = screen.getAllByTestId(
    "locationsOptionsContainerTestId"
  )[0];
  expect(locationsOptions).toHaveStyle("opacity: 0");
});
test("options are not shown when input is empty", async () => {
  const input = setup();
  const locationsOptions = screen.getAllByTestId(
    "locationsOptionsContainerTestId"
  )[0];
  fireEvent.change(input, { target: { value: "bo" } });
  expect(locationsOptions).toHaveStyle("opacity: 1");
  fireEvent.change(input, { target: { value: "" } });
  expect(locationsOptions).toHaveStyle("opacity: 0");
});
test("filtered options are shown as city/country and are chosen correct", async () => {
  const input = setup();
  fireEvent.change(input, { target: { value: "bo" } });
  const filteredLocationsList = screen.getByTestId("locationsListTestId");
  const { getAllByRole } = within(filteredLocationsList);
  const listItems = getAllByRole("listitem");
  expect(listItems.length).toEqual(1);

  listItems.forEach((item) => {
    const { queryByText } = within(item);
    expect(queryByText("Boston, USA")).toBeInTheDocument();
  });
});

test("user can select value from dropdown", async () => {
  const input = setup();
  fireEvent.change(input, { target: { value: "bo" } });
  const filteredLocationsList = screen.getByTestId("locationsListTestId");
  const { getAllByRole } = within(filteredLocationsList);
  const listItems = getAllByRole("listitem");
  fireEvent.click(listItems[0]);
  expect((input as HTMLInputElement).value).toBe("Boston, USA");
});
