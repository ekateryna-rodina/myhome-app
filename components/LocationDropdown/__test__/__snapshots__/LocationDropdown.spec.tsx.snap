// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filtered options are shown as city/country and are chosen correct: <ul>
  <li>
    Boston, USA
  </li>
</ul>
 1`] = `
<ul
  class="LocationDropdown__OptionsList-sc-q79dmt-5 hXEIkW"
  data-testid="locationsListTestId"
>
  <li>
    Boston
  </li>
</ul>
`;
